### PR TITLE
LIHADOOP-12557 Hadoop Plugin should support launching Spark scripts f…

### PR DESCRIPTION
**Pull request by nntnag17**
…rom a gateway machine.

### CODE REVIEW

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkExtension.groovy [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  /**
+   * Validates the current set of properties set on the SparkExtension. Throws an exception if the
+   * property vales cannot be validated.
+   */
+  void validateProperties() {
+    if (!zipTaskName) {
+      throw new Exception("zipTaskName must be specified by the user")
+    }
+    if (!generateTasks) {
+      throw new Exception("Method validateProperties called when generateTasks is false");
```
Comment (convexquad) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkExtension.groovy]: I don't understand this exception message - can you make it more clear?

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linkedin.gradle.spark
+
+class SparkSubmitHelper {
+
+  static final EMPTY_STRING = ""
```
Comment (convexquad) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy]: You don't need a constant for the empty string. Normally, you would use a static final member variable to declare string constants you use more than once, but for the empty string, it's ok to simply put it throughout the code. The compiler will still optimize it down to one statically stored string.

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Build spark confs
+   * @param sparkConfs The spark configurations to be passed to spark-submit
+   * @return string of the form "--conf key=value" which can be directly passed to spark-submit
+   */
+  static String buildSparkConfs(Map<String,Object> sparkConfs) {
+    if(!validate(sparkConfs)) {
+      return EMPTY_STRING;
+    }
+    return sparkConfs.collect() {
```
Comment (convexquad) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy]: I would make this a one-liner to improve readability

**Comment by convexquad** [comment]
This is really neat. People are going to love this. +1 to merge after fixing comments.

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkExtension.groovy (hunk 1 above)
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkExtension.groovy]: generateTasks is not required. I'm removing it completely.

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy (hunk 2 above)
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/spark/SparkSubmitHelper.groovy]: Done

**Pull request merged**